source/faq.txt

```diff
@@ -4,14 +4,102 @@
+
+The maximum size of each connection pool is set by the ``MaxConnectionPoolSize`` option, which
+defaults to ``100``. If the number of in-use connections to a server reaches
+the value of ``MaxConnectionPoolSize``, the next request to that server will wait
+until a connection becomes available.
+
+In addition to the sockets needed to support your application's threads,
+each ``MongoClient`` instance opens two additional sockets per server
+in your MongoDB topology for monitoring the server's state.
+For example, a client connected to a three-node replica set opens six
+monitoring sockets. If the application uses the default setting for
+``MaxConnectionPoolSize`` and only queries the primary (default) node, then
+there can be at most ``106`` total connections in the connection pool. If the
+application uses a :ref:`read preference <read-preference>` to query the
+secondary nodes, those connection pools grow and there can be
+``306`` total connections.
+
+To support high numbers of concurrent MongoDB threads
+within one process, you can increase ``MaxConnectionPoolSize``.
+
+The driver has a wait queue that limits the number of threads that can
+wait for a connection. The size of the wait queue is determined by the
+``WaitQueueMultiple`` option, which defaults to ``5``. To calculate the
+maximum wait queue size, the driver multiplies ``WaitQueueMultiple`` by
+``MaxConnectionPoolSize``. If you use the default value for each option,
+the wait queue size will be ``500``. You can also set the wait queue
+size by specifying the ``WaitQueueSize`` option, which overrides the
+other settings. However, we do not recommend changing the wait queue
+size from the default.
+
+Connection pools are rate-limited. The ``MaxConnecting`` setting
+determines the number of connections that the pool can create in
+parallel at any time. For example, if the value of ``MaxConnecting`` is
+``2``, the third thread that attempts to concurrently check out a
+connection succeeds only in one of the following cases:
+
+- One of the first two threads finishes creating a connection.
+- An existing connection is checked back into the pool.
+- The driver's ability to reuse existing connections improves due to
+  rate-limits on connection creation.
+
+You can set the minimum number of concurrent connections to
+each server by using the ``MinConnectionPoolSize`` option, which
+defaults to ``0``. The connection pool will be initialized with this
+number of sockets. If errors cause any sockets to close and the
+total number of sockets (both in-use and idle) drops below the minimum,
+the driver opens more sockets until the number reaches the minimum.
+
+You can set the maximum number of milliseconds that a connection can
+remain idle in the pool by using the ``MaxConnectionIdleTime`` option.
+Once a connection is idle for ``MaxConnectionIdleTime``, the driver
+removes and replaces it. This option defaults to 10 minutes.
```

Comment:
The driver will not replace connections that reach `MaxConnectionIdleTime` unless the pool size is less than `MinPoolSize`. This allows the driver to create more connections during bursts of activity but then slowly reap connections that are no longer required based on steady-state usage.